### ISSUE TITLE
Multi-Chain Support

### ISSUE THREAD

**Issue by danfinlay** [MetaMask/metamask-extension#1820]
MetaMask wants to support a plurality of chains in the future, but we don't want to just add a huge menu of providers, so how do we want to allow MetaMask usage with other chains?

- [ ] Make EIP 155 support optional, may require a chainId method from RPCs.
- [ ] Allow web UIs to specify the chain they use, providing an optional custom RPC provider for that chain.

Related to
#1476 
#1555

**Comment by ethernomad** [comment]
LINK also needs this: https://www.link-blockchain.org/

**Comment by chrisfranko** [comment]
The first being ETC of course? EXP was the first to submit a thorough pull request for this last year. Though I completely support the idea that it would be nice to have multi chain support its kind of crappy to again insist that ETC should be the first "of course" when it would be trivial like you said to add multiple chains at once.

**Comment by chrisfranko** [comment]
ETC being first of course. 🙄

**Comment by chrisfranko** [comment]
I dont really know why having a menu full of networks is a bad thing.. it already has a menu full of networks, but the second option to make eip155 an optional param for custom rpcs is probably the easiest to implement.

**Comment by pyskell** [comment]
> ETC being first of course. :roll_eyes:

I meant no disrespect to other networks of course and hope to see every Ethereum-based network supported.

If the ETC Cooperative were to commission this of course we'd include ETC in the first pull requests. There's also the practical aspect that we'd want to get everything working correctly/proven in the wild with one network before moving onto others for a few reasons:
* It saves on rework if we encounter bugs we didn't see during testing
* As the MetaMask team previously mentioned, they want to make sure they trust the node operators. This is something only they can determine.

We'd likely help with later PRs for ELLA and ETSC as well since those two cooperate with various ETC-related teams on other projects. Expanse, Ubiq, and all other Ethereums should have a trivially easy time getting their PRs done once all the work is complete.

Anyways, back to implementations:

For the menu I was actually wondering if it can be a two-step selection where you have an overarching menu structure like so (select outer-bullet point, then inner):
* Ethereum
    * Mainnet
    * Testnet 1
    * Testnet 2
* ETC
    * Mainnet
    * Testnet 1
* Expanse
    * Mainnet
    * Testnet 1

**Comment by chrisfranko** [comment]
Ok, sorry for getting defensive, id be willing to provide some resources to
get this done, perhaps even give an Expanse Grant to whoever wants to do
it. :) ClassicMask has already done alot of the work already for adding
network ids and stuff. Its really up to the MetaMask guys/ Ive reached out
before wanting to help and though they were very nice and respectful, they
seemed a little apprehensive of the idea. :/ Maybe our teams can get
together and just make a completely new product that is similar.

**Comment by danfinlay** [comment]
If you want ETC support, the way forward is the same as described in the parent post. All other proposals hinge on:

- Make EIP 155 support optional, may require a chainId method from RPCs.

This would be a simple enough PR to submit, we don't need to debate network menus before this has been added.

**Comment by pyskell** [comment]
@danfinlay Got it. We'll get this change submitted.

**Comment by hackmod** [comment]
Please see https://github.com/ethereum/go-ethereum/pull/15002#issuecomment-394560172

currently, `lookupNetwork()` use `net_version` value and it is the networkID.
MetaMask use this networkID as the chainID.
(https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/controllers/network/network.js#L80)

@turbobit can manage this issue (chainID != networkID cases) by hack/workaround `net_version` return value.

**Comment by hackmod** [comment]
from https://github.com/ethereum/EIPs/blob/master/EIPS/eip-155.md

(networkID column added)

NetworkID |CHAIN_ID | Chain(s)
-- | -- | --
1 | 1 | Ethereum mainnet
2 | 2 | Morden (disused)
3 | 3 | Ropsten
5 | 4 | Rinkeby
? | 30 | Rootstock mainnet
? | 31 | Rootstock testnet
? | 42 | Kovan
1 | 61 | Ethereum Classic mainnet
3 | 62 | Ethereum Classic testnet
1 | 2 | Expanse mainnet ([[1]])
1 | 8 | Ubiq mainnet ([[2]] , [[3]])
1 | 820 | Callisto mainnet 

\[1] https://github.com/expanse-org/go-expanse/blob/master/eth/config.go#L42
\[2] https://www.ethernodes.org/network/1/nodes
\[3] https://github.com/kvhnuke/etherwallet/blob/mercury/app/scripts/nodes.js


[1]: https://github.com/expanse-org/go-expanse/blob/master/eth/config.go#L42
[2]: https://www.ethernodes.org/network/1/nodes
[3]: https://github.com/kvhnuke/etherwallet/blob/mercury/app/scripts/nodes.js

**Comment by hackmod** [comment]
> [ ] Make EIP 155 support optional, may require a chainId method from RPCs.

no other web wallets even support EIP155 disabled blockchian.
without optional EIP155 support, Multi-chain support is much simpler than we expected.

**Comment by pyskell** [comment]
Can we address this? Is EIP-155 really needed if every blockchain implements it? @danfinlay

**Comment by hackmod** [comment]
See also PR #5552

**Comment by bobsummerwill** [comment]
Hey @danfinlay and @kumavis!

I am just trying to figure out where everything is sitting on this.

It looks like abortive attempts were made in 2017 and 2018 to get ETC (and other chains) working with MetaMask which we can maybe complete now without too much pain?

Everything which @pyskell said there in May 2018 remains the case.   The ETC Cooperative can drive this, find volunteers or fund developers to make the PRs, etc.

**Comment by bobsummerwill** [comment]
It looks like this change was merged and then unmerged:

https://github.com/MetaMask/metamask-extension/pull/5552

Because of this Infura issue:

https://github.com/MetaMask/metamask-extension/issues/5664

But then @kumavis did this fix:

https://github.com/MetaMask/metamask-extension/pull/5670

But this was never remerged and then everything stalled?

https://github.com/MetaMask/metamask-extension/pull/5552

**Comment by bobsummerwill** [comment]
Anyway - I am happy to jump on this horse and start riding again.   I just want to confirm where we are, and that you guys are OK with this work picking up again.

Aside - We are talking to EG at Infura too, about what it might take to get Infura ETC support.   That would be easiest of all, but may or may not be possible.

**Comment by sorpaas** [comment]
FYI, #5664 is no longer an issue -- both geth and parity added `eth_chainId` support.

**Comment by dominicletz** [comment]
Is there a API proposal for this existing that external devs could start making pull requests against? 

E.g. I could imagine a new api overload of request() with a new client only **wallet_requestChainId** method, so that each DApp can decide which ChainId to operate on.

```javascript
ethereum
  .request({ method: 'wallet_requestChainId', params: [1] })
  .then((chainId) => {
    console.log(`hexadecimal string: ${chainId}`);
    console.log(`decimal number: ${parseInt(chainId, 16)}`);
  })
  .catch((error) => {
    console.error(`Error requesting chainId: ${error.code}: ${error.message}`);
  });
```

That could then be resolved using the offical EIP-155 https://chainid.network/chains.json, or the file could also be cached as part of the metamask distribution Is this along the lines of the initial thinking? Or is there something more concrete/different in the making?

**Comment by jimthedj65** [comment]
> Is there an API proposal for this existing that external devs could start making pull requests against?
> 
> E.g. I could imagine a new api overload of request() with a new client only **wallet_requestChainId** method, so that each DApp can decide which ChainId to operate on.
> 
> ```js
> ethereum
>   .request({ method: 'wallet_requestChainId', params: [1] })
>   .then((chainId) => {
>     console.log(`hexadecimal string: ${chainId}`);
>     console.log(`decimal number: ${parseInt(chainId, 16)}`);
>   })
>   .catch((error) => {
>     console.error(`Error requesting chainId: ${error.code}: ${error.message}`);
>   });
> ```
> 
> That could then be resolved using the offical EIP-155 https://chainid.network/chains.json, or the file could also be cached as part of the metamask distribution. Is this along the lines of the initial thinking? Or is there something more concrete/different in the making?

I would seriosuly welcome this, I am building a Peering Exchange on mellanox and Intel Tech and anything that can automate Metmask at API level is a real benefit. Especially when handling highly encrypted traffic at 800 gbps plus per 2u.  This would allow me to build a central hub for automating all smart contracts based on a more flexible architecture and design.

**Comment by brad-decker** [comment]
I think, with the addition of `wallet_addEthereumChain` we can safely close this. Support multiple chains is, at the very least, part of our ongoing philosophy at MetaMask.

**Comment by danfinlay** [comment]
Thank you for that Brad, we shall close!

Anyone who's been waiting for this can [read the EIP 3086 here](https://github.com/ethereum/EIPs/issues/3086) or [read the docs here](https://docs.metamask.io/guide/rpc-api.html#wallet-addethereumchain).

**Comment by bobsummerwill** [comment]
Hurrah :-)

Thank you, MetaMask crew!

**Comment by chrisfranko** [comment]
Two years and some change in the making. Im glad to see the air around "forks" clear up.

**Comment by kelvin7891** [comment]
> MetaMask wants to support a plurality of chains in the future, but we don't want to just add a huge menu of providers, so how do we want to allow MetaMask usage with other chains?
> 
> * [ ]  Make EIP 155 support optional, may require a chainId method from RPCs.[ ]  Allow web UIs to specify the chain they use, providing an optional custom RPC provider for that chain.
> 
> Related to [#1476](https://github.com/MetaMask/metamask-extension/pull/1476) [#1555](https://github.com/MetaMask/metamask-extension/pull/1555)
